### PR TITLE
Use ashmem for shared memory for android.

### PR DESCRIPTION
**Pull request by suyambulingamrm**
Signed-off-by: suyambu.rm <suyambu.rm@samsung.com>

### CODE REVIEW

**Comment by suyambulingamrm** [comment]
This PR is to discuss using ashmem for android platform.
I thought of using ashmem() apis in shared_memory_posix, platform_shared_memory_region_posix itself.
1. If we use the original files (shared_memory_android, platform_shared_memory_region_android), we face a lot of conflict because of PlatformHandle object difference in mojo side.
2. Because if 1, I feel we need to add more handlings in the code to make the interaction between linux and android.

Please give your opinions.

**Review by hyowon — COMMENTED** [review]
First of all, I agree with your approach. The shared memory types between Android and POSIX are different:
```
#elif defined(OS_ANDROID)
  using PlatformHandle = int;
  using ScopedPlatformHandle = ScopedFD;
#else // POSIX
  using PlatformHandle = FDPair;
  using ScopedPlatformHandle = ScopedFDPair;
```

This will cause annoying code diffs. Therefore, I agree to add the ashmem related codes based on the POSIX files.

And if the browser creates a shared memory and then passes the handle to the renderer on android, we will still be opening the file in non-tmpfs without using ashmem() in CreateAnonymousSharedMemoryIfNeeded(). Could you also check this?

**Comment by younghajung** [comment]
I agree too, but read-only handle will not be supported on Android, is it OK?

**Comment by suyambulingamrm** [comment]
> First of all, I agree with your approach. The shared memory types between Android and POSIX are different:
> 
> ```
> #elif defined(OS_ANDROID)
>   using PlatformHandle = int;
>   using ScopedPlatformHandle = ScopedFD;
> #else // POSIX
>   using PlatformHandle = FDPair;
>   using ScopedPlatformHandle = ScopedFDPair;
> ```
> 
> This will cause annoying code diffs. Therefore, I agree to add the ashmem related codes based on the POSIX files.
> 
> And if the browser creates a shared memory and then passes the handle to the renderer on android, we will still be opening the file in non-tmpfs without using ashmem() in CreateAnonymousSharedMemoryIfNeeded(). Could you also check this?

You are right. I missed that. I will check on it.

**Comment by suyambulingamrm** [comment]
> I agree too, but read-only handle will not be supported on Android, is it OK?

It seem there is no separate handle for read-only case. Chromium is using ashmem_set_prot_region() api for changing the mode of the shared memory region whenever  required. We may have to further see how to handle this.

**Comment by suyambulingamrm** [comment]
Handled CreateAnonymousSharedMemoryIfNeeded() in latest patch set.

File: base/memory/shared_memory_helper.cc [hunk 1]
```diff
@@ -41,6 +45,22 @@ bool CreateAnonymousSharedMemory(const SharedMemoryCreateOptions& options,
   // It doesn't make sense to have a open-existing private piece of shmem
   DCHECK(!options.open_existing_deprecated);
 #endif  // defined(OS_LINUX)
+#if defined(ANDROID) && defined(CASTANETS)
```
Comment (hyowon) [inline base/memory/shared_memory_helper.cc]: Let's clean up the complex define branches.
Current:

```
#if !defined(OS_ANDROID) || defined(CASTANETS)
bool CreateAnonymousSharedMemory(...) {
  ...
#if defined(ANDROID) && defined(CASTANETS)
  // android specific impl.
#endif
  ...
}

bool PrepareMapFile(...) {
  // original impl.
}
#endif // !defined(OS_ANDROID) || defined(CASTANETS)
```

Below is my suggestion.

```
#if !defined(OS_ANDROID)
bool CreateAnonymousSharedMemory(...) {
  // original impl.
}

bool PrepareMapFile(...) {
  // original impl.
}
#elif defined(ANDROID) && defined(CASTANETS)
bool CreateAnonymousSharedMemory(...) {
  // only android specifc impl here.
}
#endif // !defined(OS_ANDROID)
```

File: base/memory/shared_memory_posix.cc [hunk 2]
```diff
@@ -90,11 +90,37 @@ bool SharedMemory::CreateAndMapAnonymous(size_t size) {
...
+  }
+
+  readonly_shm_ =
+      SharedMemoryHandle(FileDescriptor(dup(fd), false),
+                         options.size, shm_.GetGUID());
+  requested_size_ = options.size;
+
+  return true;
+#else
+  if (options.size == 0) return false;
```
Comment (hyowon) [inline base/memory/shared_memory_posix.cc:122]: Why did you move this code from the original line?

**Review by hyowon — CHANGES_REQUESTED** [review]

File: base/memory/shared_memory_posix.cc (hunk 2 above)
Comment (suyambulingamrm) [reply base/memory/shared_memory_posix.cc:122]: In shared_memory_android.cc, there is no validation for size == 0 case, so I moved this line below.
And it is also mentioned as below on line 296 on shared_memory_posix.cc, so I thought size = 0 while creating ashmem is valid. 

// On Android, Map can be called with a size and offset of zero to use the
// ashmem-determined size.

**Review by hyowon — APPROVED** [review]
LGTM
